File: Data Structures & Algorithms/two-integer-sum-ii/submission-2.js
class Solution {
    /**
     * @param {number[]} numbers
     * @param {number} target
     * @return {number[]}
     */
    twoSum(numbers, target) {
        let l = 0
        let r = numbers.length - 1

        while(l < r){
            const sum = numbers[r] + numbers[l]
            if(sum === target) return [l + 1, r + 1]

            if(sum > target){
                r--
            }else{
                l++
            }
        }
        return [0,0]
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-2.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        for(int i = 0; i < nums.size(); i++){
            if(i > 0 && nums[i] == nums[i - 1]) continue;
            int target = -(nums[i]);
            int l = i + 1;
            int r = nums.size() - 1;
            while(l < r){
                const int sum = nums[l] + nums[r];
                if( sum == target ){
                    result.push_back({nums[i], nums[l], nums[r]});
                    l++;
                    r--;
                    while( nums[l - 1] == nums[l]){
                        l++;
                    }
                    while( nums[r + 1] == nums[r]){
                        r--;
                    }
                }
                else if(sum > target){
                    r--;
                }else{
                    l++;
                }
            }
        }
        return result;
    }
};
